### DISCUSSION TITLE
323

### DISCUSSION THREAD

**Discussion by giscus** [Shane-Park/markdownBlog#173] [Comments]
# 323

Intro 행정안전부 소프트웨어 개발보안 가이드 https://www.mois.go.kr/frt/bbs/type001/commonSelectBoardArticle.do;jsessionid=TjAX2IwVk6hpONx8dKSZ4VTj.node10?bbsId=BBSMSTR_000000000015&nttId=88956 지난 주에 행정안전부의 소프트웨어 개발 보안 가이드를 읽다 보니 꽤나 인상적인 부분이 있었습니다. 자바의 캡슐화에 대해 단순히 private으로 변수를 생성하고 그에대해 Setter만 막아 두면, 해당 변수는 외부에서 변경이 불가능 하다고 너무나도 당연하게 여기고 있었는데 해당 문구를 보고 잠깐 고민을 해보니 자바의 캡슐화에 대해 너무 안일하게 생각하고 있었구나 싶었습니다. 캡슐화의 허점 Pub..

https://shanepark.tistory.com/323

**Comment by Firedrago95** [comment]
작성자님 고민했던 부분을 시원하게 긁어주는 포스팅이었습니다. 감사드립니다. 

1. getter메서드에서 복사하여 반환하는 경우에 getter 메서드의 단점이 완벽하게 해결된걸까요?
다른 문제가 발생할 수 있는지 궁금합니다.  클린코드 원칙에서는 getter 자체를 사용하지 않는 것을 추천하는데
복사해서 반환하는 경우라도 객체지향을 해치는 다른 문제가 있는지요

**Reply by ShanePark** [reply]
안녕하세요. getter는 무조건적인 지양은 아니고, 캡슐화에 위배되는 지나친 getter 사용을 자제하라는 뜻일거에요.
객체에서 데이터를 갖다 써야 하는 경우가 있다면 어떤 방법으로든 사용해야겠죠 그게 getter의 형태가 된 것 뿐이구요.
 본 글에서는 특정 객체가 참조중인 배열이나 컬렉션을 조회해서 활용해야 하는 경우가 있을 때 데이터를 갖다 쓸 수는 있도록 하되 원본 데이터에 영향을 주지는 않도록 (Setter의 역할은 수행하지 못하고 Getter 의 역할만 가능하도록)  하는 것에 초점을 맞췄다고 이해해주세요. 
발생하는 다른 문제가 무엇을 말씀하시는지는 제가 정확히 모르겠어서 더 자세히 답변드리진 못하지만 모든것에는 장,단점이 있기 마련이고 그 균형을 맞추는게 좋은 설계를 위한 노력이 아닐까 싶어요
